test/unit/StakingPool/calculatePremium.js

```diff
@@ -0,0 +1,352 @@
+const { expect } = require('chai');
+const { ethers } = require('hardhat');
+const { parseEther } = ethers.utils;
+const { BigNumber } = ethers;
+const { calculateSurgePremiums, calculateBasePrice, divCeil } = require('./helpers');
+const { daysToSeconds } = require('../../../lib/helpers');
+
+describe('calculatePremium', function () {
+  const stakedProductTemplate = {
+    lastEffectiveWeight: BigNumber.from(50),
+    targetWeight: BigNumber.from(70), // 70%
+    targetPrice: BigNumber.from(500), // 5%
+    nextPrice: BigNumber.from(100), // 10%
+    nextPriceUpdateTime: BigNumber.from(0),
+  };
+
+  it('should return 0 premium when period is 0', async function () {
+    const { stakingPool } = this;
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const { NXM_PER_ALLOCATION_UNIT } = this.config;
+    const stakedProduct = { ...stakedProductTemplate, nextPriceUpdateTime: timestamp };
+    const period = 0;
+    const coverAmountRaw = parseEther('100');
+    const coverAmount = divCeil(coverAmountRaw, NXM_PER_ALLOCATION_UNIT);
+
+    const totalCapacity = parseEther('100');
+    const initialCapacityUsed = BigNumber.from(0);
+
+    const { premium } = await stakingPool.calculatePremium(
+      stakedProduct,
+      period,
+      coverAmount,
+      initialCapacityUsed,
+      totalCapacity,
+      stakedProduct.targetPrice,
+      timestamp,
+    );
+    expect(premium).to.be.equal(0);
+  });
+
+  it('should overflow when calculating premium for very large capacity', async function () {
+    const { stakingPool } = this;
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const { NXM_PER_ALLOCATION_UNIT } = this.config;
+    const stakedProduct = { ...stakedProductTemplate, nextPriceUpdateTime: timestamp };
+    const period = daysToSeconds(365);
+    const coverAmountRaw = parseEther('100');
+    const coverAmount = divCeil(coverAmountRaw, NXM_PER_ALLOCATION_UNIT);
+
+    const totalCapacity = BigNumber.from(2).pow(255);
+    const initialCapacityUsed = BigNumber.from(0);
+    const basePrice = calculateBasePrice(timestamp, stakedProduct, this.config.PRICE_CHANGE_PER_DAY);
+
+    await expect(
+      stakingPool.calculatePremium(
+        stakedProduct,
+        period,
+        coverAmount,
+        initialCapacityUsed,
+        totalCapacity,
+        stakedProduct.targetPrice,
+        timestamp,
+      ),
+    ).to.be.revertedWithPanic(17); // divide-by-zero (0x11)
+    await expect(
+      stakingPool.calculatePremiumPerYear(basePrice, coverAmount, initialCapacityUsed, totalCapacity),
```

Comment:
we shouldn't test 2 functions in the same test file when dealing with unit tests.

Unit tests should normally focus on 1 function.

calculatePremiumPerYear is a different function.

If it helps to write unit tests for calculatePremiumPerYear, write them in a separate file.

I could argue it's enough to test `calculatePremium` and that would be enough.


If you want to bring an argument that calculatePremiumPerYear should be tested inidivdually as well because of complexity i could accept that.

Reply:
Ok I think it's ok to leave out premiumPerYear :ok_hand: